File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.js
function solution(s) {
    const isValidString = (str) => {
        const stack = [];
        const pairs = { ")": "(", "]": "[", "}": "{" };
        
        for (let c of str) {
            if (c in pairs) {   // 닫는 괄호
                if (!stack.length || stack.pop() !== pairs[c]) return false;
            } else {    // 여는 괄호
                stack.push(c);
            }
        }
        return stack.length === 0;
    }
    
    let answer = 0;
    let rotated = s;
    
    for (let i = 0; i < s.length; i++) {
        if (isValidString(rotated)) answer++;
        rotated = rotated.substring(1) + rotated[0];  // 문자열 회전
    }
    
    return answer;
}

File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.py
def is_correct(s):
    new_s = ""
    for c in s:
        if new_s and ((new_s[-1] == "(" and c == ")") or (new_s[-1] == "[" and c == "]") or (new_s[-1] == "{" and c == "}")):
            new_s = new_s[:-1]
        else:
            new_s += c
    
    return True if not new_s else False

def solution(s):
    answer = 0
    
    for _ in range(len(s)):        
        if is_correct(s):
            answer += 1
            
        s = s[1:] + s[0]
    
    return answer 